src/core/Akka/IO/TcpListener.cs

```diff
@@ -40,79 +40,136 @@ partial class TcpListener : ActorBase, IRequiresMessageQueue<IUnboundedMessageQu
         {
             switch (message)
             {
                 case SocketEvent evt:
                     var saea = evt.Args;
                     if (saea.SocketError == SocketError.Success)
                         Context.ActorOf(Props.Create<TcpIncomingConnection>(_tcp, saea.AcceptSocket, _bind.Handler, _bind.Options, _bind.PullMode).WithDeploy(Deploy.Local));
+                    
                     saea.AcceptSocket = null;
-
                     if (!_socket.AcceptAsync(saea))
                         Self.Tell(new SocketEvent(saea));
                     return true;
 
                 case Tcp.ResumeAccepting resumeAccepting:
                     _acceptLimit = resumeAccepting.BatchSize;
+                    // TODO: this is dangerous, previous async args are not disposed and there's no guarantee that they're not still receiving data
                     _saeas = Accept(_acceptLimit).ToArray();
                     return true;
 
                 case Tcp.Unbind _:
-                    _log.Debug("Unbinding endpoint {0}", _bind.LocalAddress);
-                    _socket.Dispose();
-                    Sender.Tell(Tcp.Unbound.Instance);
+                    Become(Unbinding(Sender));
+                    UnbindAsync().PipeTo(Self);
+                    return true;
+
+                default:
+                    return false;
+            }
+        };
+
+        private Receive Unbinding(IActorRef requester) => message =>
+        {
+            switch (message)
+            {
+                case Tcp.Unbound unbound:
+                    requester.Tell(unbound);
                     _log.Debug("Unbound endpoint {0}, stopping listener", _bind.LocalAddress);
                     Context.Stop(Self);
                     return true;
-
+                
+                case Status.Failure fail:
+                    _log.Error(fail.Cause, "Failed to unbind TCP listener for address [{0}]", _bind.LocalAddress);
+                    Context.Stop(Self);
+                    return true;
+                
                 default:
                     return false;
             }
+        };
+        
+        private IEnumerable<SocketAsyncEventArgs> Accept(int limit)
+        {
+            for(var i = 0; i < limit; i++)
+            {
+                var self = Self;
+                var saea = new SocketAsyncEventArgs();
+                saea.Completed += (s, e) => self.Tell(new SocketEvent(e));
+                if (!_socket.AcceptAsync(saea))
+                    Self.Tell(new SocketEvent(saea));
+                yield return saea;
+            }
+        }
+
+        private async Task<Tcp.Bound> BindAsync()
+        {
+            _socket = new Socket(_bind.LocalAddress.AddressFamily, SocketType.Stream, ProtocolType.Tcp) { Blocking = false };
+                
+            _bind.Options.ForEach(x => x.BeforeServerSocketBind(_socket));
+            _socket.Bind(_bind.LocalAddress);
```

Comment:
No `await`-able methods in .NET Standard 2.0?

Reply:
No, `Bind` and `Listen` are blocking